src/llama.cpp

```diff
@@ -5368,13 +5401,20 @@ static void llm_load_hparams(
                         // granite uses a vocab with len 49152
                         case 32: model.type = hparams.n_vocab == 49152 ? e_model::MODEL_3B : (hparams.n_vocab < 40000 ? e_model::MODEL_7B : e_model::MODEL_8B); break;
                         case 36: model.type = e_model::MODEL_8B; break; // granite
-                        case 40: model.type = e_model::MODEL_13B; break;
+                        case 40: model.type = (hparams.n_vocab == 49152 || hparams.n_vocab == 49156) ? e_model::MODEL_3B : e_model::MODEL_13B; break;
```

Comment:
Here, I think it might be more appropriate to use `hparams.n_embd` to differentiate the sizes. `hparams.n_vocab` was used to distinguish between 7B and 8B Llama models because that was the only difference, but between a 3B and a 13B model with the same layer count, I'm sure `n_embd` isn't the same.

```suggestion
                        case 40: model.type = (hparams.n_embd == 2304) ? e_model::MODEL_3B : e_model::MODEL_13B; break;
```

Although maybe this should be a nested switch-case instead.

Reply:
I took the other suggestion to split this whole switch clause out, so this got quite a bit simpler. Thanks!